### PR TITLE
Tbmr features (purely topological adaptation on MSER)

### PR DESCRIPTION
**Pull request by crownedone**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/817

Some questions:
- Generalized Component tree calculation could be an additional feature, instead of hiding in the implementation of TBMR (as it can be used by MSER, TBMR, ...). I could do an extra module for that (where)? 
- Is this the right place for TBMR? (In xFeatures2d)

Thanks for your work!

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test
- [x] The feature is well documented.
- [ ] and sample code can be built with the project CMake

### CODE REVIEW

**Comment by crownedone** [comment]
I do not understand, why the builds fail. Its related to python wrapper. Help appreciated!

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 1]
```diff
@@ -70,25 +72,27 @@ namespace xfeatures2d
...
-computed by efficiently comparing image intensities over a retinal sampling pattern. FREAKs are in
-general faster to compute with lower memory load and also more robust than SIFT, SURF or BRISK.
-They are competitive alternatives to existing keypoints in particular for embedded applications.
+The algorithm propose a novel keypoint descriptor inspired by the human visual
+system and more precisely the retina, coined Fast Retina Key- point (FREAK). A
+cascade of binary strings is computed by efficiently comparing image intensities
+over a retinal sampling pattern. FREAKs are in general faster to compute with
+lower memory load and also more robust than SIFT, SURF or BRISK. They are
+competitive alternatives to existing keypoints in particular for embedded
+applications.
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:84]: Need to remove massive unrelated changes from this PR.

**Review by alalek — CHANGES_REQUESTED** [review]
Keep patch clean and minimal for reviewing.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 1 above)
Comment (crownedone) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:84]: Is there a .clang format file that is conform with opencv indentation etc.?
Else I can do it by hand.

Comment (alalek) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:84]: There is no clang format file (anyway it can be used with new files only).

Comment (crownedone) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: ok, that should be done.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 2]
```diff
@@ -945,6 +945,40 @@ class CV_EXPORTS AffineFeature2D : public Feature2D
...
+uses a Morse-theory approach to extract features.
+
+Features are ellipses (similar to MSER, however a MSER feature can never be a
+TBMR feature and vice versa).
+
+*/
+class CV_EXPORTS TBMR : public AffineFeature2D
+{
+public:
+    CV_WRAP static Ptr<TBMR> create(int min_area = 60,
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:968]: 1. Class itself should be wrapped: `class CV_EXPORTS` => `class CV_EXPORTS_W`
2. Base class should be wrapped too: `class CV_EXPORTS_W AffineFeature2D`

File: modules/xfeatures2d/src/tbmr.cpp [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+        // topology is above its parent, if not, discard this critical
+        // node.
+        //
+        // note also that we do not select the critical nodes themselves
+        // as final TBMRs
+        //--------------------------------------------------------------------------
+
+        uint *numSons = (uint *)calloc(imSize, sizeof(uint));
+        uint vecNodesSize = imaAttribute[S_ptr[0]][0];               // area
+        uint *vecNodes = (uint *)calloc(vecNodesSize, sizeof(uint)); // area
```
Comment (alalek) [inline modules/xfeatures2d/src/tbmr.cpp:299]: > malloc/calloc/free/new/delete

Avoid using of manual memory management (it is not exception safe at least).
Consider using `cv::AutoBuffer` or `std::vector` instead.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comments below.

File: modules/xfeatures2d/src/tbmr.cpp (hunk 3 above)
Comment (crownedone) [reply modules/xfeatures2d/src/tbmr.cpp]: alright, done.

File: modules/xfeatures2d/src/tbmr.cpp [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+{
+namespace xfeatures2d
+{
+/*!
+MSD Image Pyramid. (from msd.cpp)
+*/
+class TBMRImagePyramid
+{
+    // Multi-threaded construction of the scale-space pyramid
+    struct TBMRImagePyramidBuilder : ParallelLoopBody
```
[the remark this answers is not in the document]
Comment (crownedone) [reply modules/xfeatures2d/src/tbmr.cpp]: good idea! done.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 2 above)
Comment (crownedone) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:968]: ok, done

File: modules/xfeatures2d/src/msd_pyramid.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
+///////////// see LICENSE.txt in the OpenCV root directory //////////////
```
Comment (alalek) [inline modules/xfeatures2d/src/msd_pyramid.hpp:1]: Please use short license header: https://github.com/opencv/opencv/wiki/Coding_Style_Guide#file-structure

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

File: modules/xfeatures2d/test/test_features2d.cpp [hunk 6]
```diff
@@ -85,6 +85,12 @@ TEST( Features2d_Detector_Harris_Laplace_Affine, regression )
     test.safe_run();
 }
 
+TEST(Features2d_Detector_TBMR_Affine, regression)
+{
+    CV_FeatureDetectorTest test("detector-tbmr-affine", TBMR::create());
```
Comment (alalek) [inline modules/xfeatures2d/test/test_features2d.cpp:90]: Please prepare PR with test data into `opencv_extra` repository (using the same source branch name).
To receive accurate result please use:
- "Debug" build with disabled optimizations: `cmake -DCV_DISABLE_OPTIMIZATION=ON ...` (please run this in new (or clean) build directory, because CMake caches some values like `WITH_IPP`).
- disable multi-threading: `./bin/opencv_test_xfeatures2d --gtest_filter=*TBMR* --test_threads=1`

**Review by alalek — COMMENTED** [review]
Well done. Thank you for updates!

File: modules/xfeatures2d/test/test_features2d.cpp (hunk 6 above)
Comment (crownedone) [reply modules/xfeatures2d/test/test_features2d.cpp:90]: with a PR in opencv_extra as well?

File: modules/xfeatures2d/src/msd_pyramid.hpp (hunk 5 above)
Comment (crownedone) [reply modules/xfeatures2d/src/msd_pyramid.hpp]: added.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 7]
```diff
@@ -945,6 +945,40 @@ class CV_EXPORTS AffineFeature2D : public Feature2D
...
+
+Features are ellipses (similar to MSER, however a MSER feature can never be a
+TBMR feature and vice versa).
+
+*/
+class CV_EXPORTS_W TBMR : public AffineFeature2D
+{
+public:
+    CV_WRAP static Ptr<TBMR> create(int min_area = 60,
+        float max_area_relative = 0.01,
```
[the remark this answers is not in the document]
Comment (crownedone) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: fixed

File: modules/xfeatures2d/test/test_features2d.cpp (hunk 6 above)
Comment (alalek) [reply modules/xfeatures2d/test/test_features2d.cpp:90]: Yes, please open PR in `opencv_extra` for reviewing and merging.

Comment (crownedone) [reply modules/xfeatures2d/test/test_features2d.cpp:90]: ok there it is: [TBMR features (testdata)](https://github.com/opencv/opencv_extra/pull/817)

**Review by alalek — APPROVED** [review]
Well done!

**Comment by crownedone** [comment]
Thanks!
